contrib/http_base/src/main/java/io/opentelemetry/contrib/http/core/AbstractHttpHandler.java

```diff
@@ -0,0 +1,204 @@
+    } else {
+      this.statusConverter = statusConverter;
+    }
+    if (meter == null) {
+      this.meter = OpenTelemetry.getMeterFactory().get(INSTRUMENTATION_LIB_ID);
+    } else {
+      this.meter = meter;
+    }
+  }
+
+  /**
+   * A convenience to record a {@link Event} with given parameters.
+   *
+   * @param span the span which this {@code Event} will be added to.
+   * @param type the type of event.
+   * @param uncompressedMessageSize size of the message before compressed (optional).
+   * @param compressedMessageSize size of the message after compressed (optional).
+   */
+  static void recordMessageEvent(
+      Span span,
+      long id,
+      AttributeValue type,
+      long uncompressedMessageSize,
+      long compressedMessageSize) {
+    Map<String, AttributeValue> attributes = new HashMap<>();
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_TYPE, type);
+    attributes.put(HttpTraceConstants.MSG_EVENT_ATTR_ID, AttributeValue.longAttributeValue(id));
+    if (compressedMessageSize > 0L) {
+      attributes.put(
+          HttpTraceConstants.MSG_EVENT_ATTR_COMPRESSED,
+          AttributeValue.longAttributeValue(compressedMessageSize));
+    }
+    if (uncompressedMessageSize > 0L) {
+      attributes.put(
+          HttpTraceConstants.MSG_EVENT_ATTR_UNCOMPRESSED,
+          AttributeValue.longAttributeValue(uncompressedMessageSize));
+    }
+    span.addEvent(HttpTraceConstants.MSG_EVENT_NAME, attributes);
+  }
+
+  private static void putAttributeIfNotEmptyOrNull(Span span, String key, @Nullable String value) {
+    if (value != null && !value.isEmpty()) {
+      span.setAttribute(key, AttributeValue.stringAttributeValue(value));
+    }
+  }
+
+  /**
+   * Instrument an HTTP span after a message is sent. Typically called for every chunk of request or
+   * response is sent.
+   *
+   * @param context request specific {@link HttpRequestContext}
+   * @param bytes bytes sent.
+   */
+  public final void handleMessageSent(HttpRequestContext context, long bytes) {
+    checkNotNull(context, "context is required");
+    context.sentMessageSize.addAndGet(bytes);
+    recordMessageEvent(
+        context.span,
+        context.sentSeqId.addAndGet(1L),
+        HttpTraceConstants.MSG_EVENT_ATTR_SENT,
+        bytes,
+        0L);
+  }
+
+  /**
+   * Instrument an HTTP span after a message is received. Typically called for every chunk of
+   * request or response is received.
+   *
+   * @param context request specific {@link HttpRequestContext}
+   * @param bytes bytes received.
+   */
+  public final void handleMessageReceived(HttpRequestContext context, long bytes) {
+    checkNotNull(context, "context is required");
+    context.receiveMessageSize.addAndGet(bytes);
+    recordMessageEvent(
+        context.span,
+        context.receviedSeqId.addAndGet(1L),
+        HttpTraceConstants.MSG_EVENT_ATTR_RECEIVED,
+        bytes,
+        0L);
+  }
+
+  void spanEnd(Span span, int httpStatus, @Nullable Throwable error) {
```

Comment:
can we call this `endSpan` so the verb is first?

Reply:
Fixed